src/fuel-vm/index.md

```diff
@@ -95,54 +95,39 @@ Then the following registers are initialized (without explicit initialization, a
 
-## Predicate Estimation
+### Predicate Estimation and Verification
 
-For any input of type [`InputType.Coin`](../tx-format/index.md) or [`InputType.Message`](../tx-format/index.md), a non-zero `predicateLength` field means the UTXO being spent is a [`P2SH`](https://en.bitcoin.it/wiki/P2SH) rather than a [`P2PKH`](https://en.bitcoin.it/P2PKH) output.
+There are two ways to run predicates on the VM:
 
-For each such input in the transaction, the VM is [initialized](#vm-initialization), then:
-
-1. `$pc` and `$is` are set to the start of the input's `predicate` field.
-1. `$ggas` and `$cgas` are set to `MAX_GAS_PER_PREDICATE`.
-
-Predicate estimation will fail if gas is exhausted during execution.
-
-During predicate mode, hitting any of the following instructions causes predicate estimation to halt, returning Boolean `false`:
-
-1. Any [contract instruction](./instruction-set.md#contract-instructions).
-
-In addition, during predicate mode if `$pc` is set to a value greater than the end of predicate bytecode (this would allow bytecode outside the actual predicate), predicate estimation halts returning Boolean `false`.
-
-A predicate that halts without returning Boolean `true` would not pass verification, making the entire transaction invalid. Note that predicate validity is monotonic with respect to time (i.e. if a predicate evaluates to `true` then it will always evaluate to `true` in the future).
-
-After successful execution, `predicateGasUsed` is set to `MAX_GAS_PER_PREDICATE - $ggas`.
-
-## Predicate Verification
+1. Estimation: runs the predicate and updates the amount of gas used
+1. Verification: runs the predicate and verifies the amount of gas used matches the input
 
 For any input of type [`InputType.Coin`](../tx-format/input.md#inputcoin) or [`InputType.Message`](../tx-format/input.md#inputmessage), a non-zero `predicateLength` field means the UTXO being spent is a [`P2SH`](https://en.bitcoin.it/P2SH) rather than a [`P2PKH`](https://en.bitcoin.it/P2PKH) output.
 
 For each such input in the transaction, the VM is [initialized](#vm-initialization), then:
 
 1. `$pc` and `$is` are set to the start of the input's `predicate` field.
-1. `$ggas` and `$cgas` are set to `predicateGasUsed`.
+1. `$ggas` and `$cgas` are set to `MAX_GAS_PER_PREDICATE` for estimation, and `predicateGasUsed` for verification.
 
-Predicate verification will fail if gas is exhausted during execution.
+Predicate execution will fail if gas is exhausted during execution.
 
-During predicate mode, hitting any [contract instruction](./instruction-set.md#contract-instructions) causes predicate verification to halt, returning Boolean `false`.
+During predicate mode, hitting any [contract instruction](./instruction-set.md#contract-instructions) (except `ldc` with non-contract target) causes predicate verification to halt, returning Boolean `false`.
```

Comment:
Hmm, it looks like before we should allow usage of blob instructions in the VM as well(because they are not contract instruction)=D So now it works as expected